### ISSUE TITLE
Bazel Integration

### ISSUE THREAD

**Issue by dmadisetti** [google/autocxx#478]
[`rules_rust`](https://github.com/bazelbuild/rules_rust) is the canonical bazel integration for rust. It looks like it would be pretty easy to patch something together with [`cargo_build_script`](https://bazelbuild.github.io/rules_rust/cargo_build_script.html), but keeping paths relative during build looks like it might be a pain.

Maybe it's easier than it looks? It'd just be nice to have a canonical example to reference, or a starlark rule I can blindlessly call.

**Comment by adetaylor** [comment]
@martinboehme do you have any advice you can offer here?

**Comment by martinboehme** [comment]
If you want to do it right, it's a bit more involved, because you need to use `cc_common` to get the right command line / environment variables to pass on to the C++ compiler.

I've got a proof of concept that does this in Starlark, and I've been considering submitting it to rules_rust or autocxx, but it still needs some cleaning up and some tests before it's ready for that.

rules_rust is the most central place, but I think it only contains rules for relatively stable / canonical components. autocxx is still pretty experimental, so maybe the autocxx repo would be a better place for the rule to live, at least for now. @adetaylor , what do you think?

**Comment by adetaylor** [comment]
I'm certainly open to having bazel rules in the autocxx repo, so long as they are tested in github actions - hopefully much of that can be copied from https://github.com/dtolnay/cxx/blob/master/.github/workflows/ci.yml?

**Comment by adetaylor** [comment]
I'm going to close this for now, but I'm certainly happy to accept a pull request for bazel rules in future (on the condition that it's automatically tested on github CI - this is something that has been achieved for cxx, so it's obviously possible)

**Comment by jon-chuang** [comment]
I would be quite interested in implementing the autocxx equivalent of cxx's cxxbridge-cmd tool, which dumps the autogenerated C++ headers that can then be consumed by Bazel BUILD. 

The use case is leveraging existing Bazel BUILD system in a large C++ codebase.

The pathway I am thinking of for packaging the build in cargo is to first download and install `autocxxx-bridge`, run `autocxxbridge-cmd` to dump the headers and cc files, then invoke bazel from the cargo build script via OS env cmdline to build the shared object and archive files.

As a matter of fact I am not even sure if this pathway is needed, since one is very likely autogenerating rust-side code more than C++ for this use case. So do let me know if so, and one can skip the step of needing to build a `autocxxbridge-cmd` tool.

**Comment by jon-chuang** [comment]
Actually, thinking a bit further, this sort of defeats the purpose. The purpose of using bazel is to first resolve, download and build external C++ dependencies. Since neither autocxx nor cxx can function without those dependencies first available, I'm not sure how Bazel can be integrated with cxx/autocxx for them to do their work,

**Comment by adetaylor** [comment]
Hi, the equivalent of `cxxbridge-cmd` exists, it's https://crates.io/crates/autocxx-gen/, and it's intended for exactly the same purposes of embedding into build systems like Bazel.

You use it exactly the same way as `cxxbridge-cmd` except that it also needs to know the C++ include path and any `#defines`. This is _massively_ annoying to get out of our build system, `gn` (which is very similar to Bazel in its need for static dependencies everywhere). We had to create a fake C++ toolchain which, instead of having a compiler, ran a python script that recorded the include path and any applicable definitions, such that they could later be retrieved by the script which called `autocxx_gen`. Hopefully it won't be quite so annoying for Bazel.

Good luck achieving the same for Bazel and I'm happy to answer any questions!

**Comment by jon-chuang** [comment]
Thanks @adetaylor ! Is this or the tool that leverages the data from `cc_common` that @martinboehme alluded to available somewhere? I'd like to see if I can make it work for me and if so, perhaps later on it could be cleaned up for general use, living in this repo.

**Comment by jon-chuang** [comment]
Actually, it was thankfully simple to get the include dirs from a bazel rule:
`print(ctx.attr.deps[0][CcInfo].compilation_context.includes.to_list())`

**Comment by bsilver8192** [comment]
Leaving a note here that I'm working on this. I got a simple example working today, going to keep working on it. I'll share a link here when I've got something more fully working.

**Comment by bsilver8192** [comment]
I've got something that kinda-sorta works. However, I ran into an interesting problem: Bazel toolchains come with flags for _a_ C++ compiler, not necessarily the same as the libclang that autocxx passes flags to. I'm currently using a libclang from most of my C++ toolchains, and the other C++ toolchain is a gcc of similar vintage that accepts most of the same flags, but I'm not sure how to solve this in general. Some ideas:
1. Require using a matching libclang and Clang-based C++ toolchain.
2. Allowlist and map recognized flags. This is hard to maintain.
3. Blocklist and map known-bad flags. This is also hard to maintain.
4. Require the caller to map flags. This just punts the problem to the user, which isn't very helpful. Maybe some kind of API that allows choosing and extending any of the other options?

Warning flags are also interesting. My toolchain enables `-Wunused-private-field`, which somehow broke things when it triggered. Maybe the `-Werror` is taking effect, but only the warnings get printed for some reason? I think giving the user a way to disable any problematic warnings they have is fine, it's hard to do anything else. Or maybe just filter out `-W*` flags, although that means they can't be disable either.

Anybody have thoughts on how to handle this?
